### PR TITLE
feat: initial implementation

### PR DESCRIPTION
**Pull request by andrewrynhard**
Signed-off-by: Andrew Rynhard <andrew@rynhard.io>

### CODE REVIEW

File: api/types/api.proto [hunk 1]
```diff
@@ -0,0 +1,26 @@
+syntax = "proto3";
```
Comment (andrewrynhard) [inline api/types/api.proto:1]: Should we have a distinction between state APIs and observability APIs? Whether that be a naming convention, separate files, or both?

File: docs/spec/3-Spec.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
+# Spec
+
+## Technologies
+
+- eBPF
+- Protobuf
+- gRPC
+
+## Eventing
```
Comment (andrewrynhard) [inline docs/spec/3-Spec.md]: Rename to "State"

File: main.go [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+type Options struct {
+	Address string
+	Plugins string
+}
+
+var options Options
+
+func init() {
+	flag.StringVar(&options.Plugins, "plugins-dir", "/lib/engine", "the directory to load plugins from")
+	flag.StringVar(&options.Address, "grpc-address", "", "the gRPC address to connect to")
```
Comment (andrewrynhard) [inline main.go]: This should be `runtime-address`. We should enforce that only one runtime is used at a time.

File: internal/aggregator/v1alpha1/v1alpha_aggregator.go [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+		err := source.Load()
+		if err != nil {
+			return err
+		}
+
+		go func() {
+			for msg := range source.Recv() {
+				for _, sink := range a.sinks {
+					go sink.Send(msg)
```
Comment (smira) [inline internal/aggregator/v1alpha1/v1alpha_aggregator.go]: feels like unbound goroutine creation... this could be a problem

Comment (andrewrynhard) [reply internal/aggregator/v1alpha1/v1alpha_aggregator.go]: I plan on getting the event system you wrote for Talos and replacing this whole implementation. This  whole aggregator thing was thrown together for the poc. Actually, that is what I wanted to chat with you about today.

Comment (andrewrynhard) [reply internal/aggregator/v1alpha1/v1alpha_aggregator.go]: Updated to use the event interface from Talos.

File: api/types/api.proto [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+package api;
+
+option go_package = "github.com/talos-systems/os-engine/api/types";
+
+import "google/protobuf/any.proto";
+import "google/protobuf/empty.proto";
+
+service Sink { rpc Events(stream Event) returns (google.protobuf.Empty); }
```
Comment (andrewrynhard) [inline api/types/api.proto:10]: I still need to think through, and implement an example of a gRPC source.

File: examples/grpc_sink/main.go [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+	"os"
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/talos-systems/os-engine/api/types"
+	"google.golang.org/grpc"
+)
+
+type server struct{}
+
+func (s *server) Events(stream types.Sink_EventsServer) error {
```
Comment (andrewrynhard) [inline examples/grpc_sink/main.go]: Needs comments to document that this implements the higher level gRPC interface, while the gRPC sink loader implements the internal sink interface.

File: examples/go_plugin_sink/main.go [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+var _ sink.Sink = (*example)(nil)
+
+type example struct {
+	events.Streamer
+}
+
+var Sink example
+
+func (e *example) Load(stream events.Streamer) error {
```
Comment (andrewrynhard) [inline examples/go_plugin_sink/main.go:18]: Needs comment documenting that this implements the internal sink interface.

File: internal/events/events.go [hunk 8]
```diff
@@ -0,0 +1,326 @@
+package events
```
Comment (andrewrynhard) [inline internal/events/events.go]: This still has Talos nomenclature in comments. Remove references to "runtime" and "v1alpha1."

File: api/machinery/constants.go [hunk 9]
```diff
@@ -0,0 +1,16 @@
+package machinery
+
+const (
+	DomainStorage = "storage"
+)
```
Comment (andrewrynhard) [inline api/machinery/constants.go:5]: This should be an iota.

File: README.md [hunk 10]
```diff
@@ -1 +1,12 @@
-# os-engine
\ No newline at end of file
+# os-engine
```
Comment (alejandrox1) [inline README.md:1]: How would you describe os-engine? whats the goal?

File: docs/spec/1-Goals.md [hunk 11]
```diff
@@ -0,0 +1,10 @@
+# Goals
+
+## State
```
Comment (alejandrox1) [inline docs/spec/1-Goals.md]: What is state in this context?

File: docs/spec/1-Goals.md [hunk 12]
```diff
@@ -0,0 +1,10 @@
...
+
+- Interfaces
+- Block devices
+
+## Observability
```
Comment (alejandrox1) [inline docs/spec/1-Goals.md:12]: What does observability mean in the context of os engine?

File: docs/spec/2-Concepts.md [hunk 13]
```diff
@@ -0,0 +1,31 @@
+# Concepts
+
+## Source
+
+A source is an event generator.
```
Comment (alejandrox1) [inline docs/spec/2-Concepts.md:9]: How does a source relate to state (from the goals doc)?

File: docs/spec/2-Concepts.md [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+
+### Domain
+
+A domain is a classification of an event based on factors such as
+
+- The Linux subsytem from which the event is generated
+
+## Sink
+
+A sink is a consumer of events.
```
Comment (alejandrox1) [inline docs/spec/2-Concepts.md]: nit: maybe move this section in between "source" and "loader".

File: docs/spec/2-Concepts.md [hunk 15]
```diff
@@ -0,0 +1,31 @@
+# Concepts
+
+## Source
+
+A source is an event generator.
+
+## Loader
+
+A loader is anything that can load a source, or sink into the engine.
```
Comment (alejandrox1) [inline docs/spec/2-Concepts.md]: What does it mean to "load" something?

File: docs/spec/2-Concepts.md [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+
+## Event
+
+An event is a data structure that represents a kernel event.
```
Comment (alejandrox1) [inline docs/spec/2-Concepts.md]: nit: this should probably be moved before "source".

File: main.go [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+	Runtime string
+	Plugins string
+	Socket  string
+	Debug   bool
+}
+
+var options Options
+
+func init() {
+	flag.StringVar(&options.Runtime, "runtime", "/lib/engine/runtime/runtime.so", "the path to the runtime")
```
Comment (alejandrox1) [inline main.go:25]: What would this be?
pkg/runtime points to this being an interface that implements a sink but still not super clear on what `/lib/engine/runtime/runtime.so` could be.

**Review by alejandrox1 — COMMENTED** [review]
Mostly documentation comments- trying to wrap my head around the design.

File: docs/spec/2-Concepts.md (hunk 13 above)
Comment (andrewrynhard) [reply docs/spec/2-Concepts.md:9]: Updating this. Looks like we will define state in os-runtime.

File: main.go (hunk 17 above)
Comment (andrewrynhard) [reply main.go:25]: :thinking: Still working on how engine and runtime work together (and if they even should be separate). I think I can address this in the "concepts" doc.

**Review by alejandrox1 — COMMENTED** [review]
lgtm

**Pull request merged**